Flush KML to disk ~60s when moving

tommywatson opened the pull request:
Flush the KML file to disk periodically in case the app crashes or power fails.
I'm not that up on KML and it looks like there is no time stamp in the spec for the coordinate tag, so if the app does crash/run out of bat, the user is really left with 1/2 the info which, while slightly better than a 0 length file, it isn't that great...
I'm curious, do we need these coordinates? Or could we just write the new points placemarks out every minute or so?
Could we write the points first and put this placemark at the bottom of the file? It seems like a bung file would be slightly more useful if it was points first...

JohnW4SBA commented:
> looks like there is no time stamp in the spec

I thought there was, and that I'd seen it in the dim past when looking 
at a file from the infamous Y2k vintage (~2003?) Lowrance 1000 (aka L1k).

> crash/run out of bat, the user is really left with 1/2 the info

Somehow, my L1k creates complete tracks that near as I can tell are 
within a mile of unit on/off at 125ktas.  I have Tracks permanently 
turned on in L1k Settings.  So after engine start I just turn it on. 
Once it groggily awakes and tediously (by contemporary standards) 
acquires GPS, it starts recording points to internal memory.  Now that 
we have at least four other GPS receivers aboard I generally fly with 
the battery backup disabled (paper at the end of 1 of the four AA 
batteries).  So it just keeps recording Tracks until I shut down the 
engine and simply turn off the master.  Somehow it also saves to a max 
point limit (2k as I recall) and then FIFO, deleting the oldest points 
to maintain the limit number.  To use the files I have to remove the SD 
chip from the L1k and copy it to a laptop or other device, then convert 
it on the website I mentioned before.  Maybe our Android devices are 
crippled by Google in some way so that we don't have access to the raw 
GPS data the L1k is using?

If it would be helpful I could send you a raw data file from the L1k, 
and/or the KML file created by the GPSvis website http://gpsvisualizer.com/

tommywatson commented:
This is what I found;

https://developers.google.com/kml/documentation/kmlreference#coordinates
<coordinates>(required)
Four or more tuples, each consisting of floating point values for longitude, latitude, and altitude. The altitude component is optional. Do not include spaces within a tuple. The last coordinate must be the same as the first coordinate. Coordinates are expressed in decimal degrees only.

I really don't know, we are creating a <LineString> element so I assume it applies, I was just hoping we could put it at the end of the file, so we would end up with the position elements on failure rather than the coordinates that don't have timestamp/speed information. It appears that this works, I modified the attached file and moved the coordinates after the position reports and it can be parsed by software on the web... ie. http://www.gpsvisualizer.com/

If it's okay I'll go ahead and change the KMLRecorder to write the positions as they come in and close the file with the Avare Flight Path element.

[2016-10-01_11-43-29.KML.txt](https://github.com/apps4av/avare/files/507568/2016-10-01_11-43-29.KML.txt)

apps4av commented:
Test report:
Short test and I noticed that the acknowledgement / notice that the "..kml has been saved" has disappeared, when the Tracks off  state changes / toggled. In addition when I forced a Power off in the middle of the flight, after I rebooted / restarted, all my Data / Charts... came up with an ?.....
I am not sure what the this "improvement" is but at this point it has more problems, and I see No improvement. I did a lot of testing in the old code and the .kml portion worked very reliably, it had some issues but those still exist, and in addition the new code is affecting the Data in some strange way.

tommywatson commented:
All this does is flush the file to "disk" every time a point is logged > 60s since the last time a flush occurred. If the power went out before just a 0 length file was left behind, see: https://groups.google.com/d/msg/apps4av-forum/tSn3NO7E56I/FeCiEtksAwAJ
This change will at least have most of the data though I believe it's only barely better than the 0 length file as no time stamp is in the 'coordinates' tag, so you will get a track but not speed, see my questions above.

1) "kml has been saved" - do you have Flight Tracking -> Automatic Track Post set to None? If this is set to send email or send to application you do not see the kml saved toast, this is legacy code behavior, see: https://github.com/apps4av/avare/blob/master/app/src/main/java/com/ds/avare/LocationActivity.java#L996 only the do nothing option makes the toast. My belief is you were testing in an emulator and one of the other options is set, which fails due to configuration.
2) I cannot explain the "charts ?" issue, this didn't happen when I tested as described on real hardware or an emulator. Can you reproduce? Is this the only difference between master? Or are there other changes in your build?

Holding open file handles (which we are doing) in any operating system is dangerous when you power off can lead to corruption. What I'd really like to do is write out a complete log periodically rather than hold the handle open indefinitely until the app closes, that way we will always have a decent track on failure, I'm just getting my feet wet with these commits before I commit to the time to fix the logging.

tommywatson commented:
Better solution implemented
